pkg/operator/observe_config.go

```diff
@@ -149,13 +154,172 @@ func observeInternalRegistryHostname(listers Listers, observedConfig map[string]
+	if err = observeField(observedConfig, nodeSelector.MatchLabels, "build.buildOverrides.nodeSelector"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildOverrides.nodeSelector", err)
+	}
+	// Control plane config does not support MatchExpressions yet
+	if len(nodeSelector.MatchExpressions) > 0 {
+		return nil, fmt.Errorf("config.Build: %s is not supported", "buildOverrides.nodeSelector.matchExpressions")
+	}
+	if err = observeField(observedConfig, build.Spec.BuildOverrides.Tolerations, "build.buildOverrides.tolerations"); err != nil {
+		return nil, fmt.Errorf("failed to observe %s: %v", "build.buildOverrides.tolerations", err)
+	}
+
+	// TODO: 1) generate CA bundle from ConfigMapRef
+	//       2) write CA bundle to a ConfigMap in the controller-manager's namespace
+	//       3) wire in logic to force a rollout if CA bundle changes
+	// additionalCA := build.Spec.AdditionalTrustedCA
+	// if len(additionalCA.Name) > 0 {
+	// 	unstructured.SetNestedField(observedConfig, "/var/run/openshift.io/config/certs/additional-ca.crt", "build", "additionalTrustedCA")
+	// }
+	return observedConfig, nil
+}
+
+// observeField sets the nested fieldName's value in the provided observedConfig.
+// If the provided value is nil or empty, no value is set.
+func observeField(observedConfig map[string]interface{}, val interface{}, fieldName string) error {
+	nestedFields := strings.Split(fieldName, ".")
+	if val == nil {
+		return nil
+	}
+	var err error
+	switch v := val.(type) {
+	case int64, bool:
+		err = unstructured.SetNestedField(observedConfig, v, nestedFields...)
+	case string:
+		if len(v) > 0 {
+			err = unstructured.SetNestedField(observedConfig, v, nestedFields...)
+		}
+	case []corev1.EnvVar:
+		if len(v) > 0 {
+			data, err := envVarsToRawEnvVars(v)
+			if err != nil {
+				return err
+			}
+			err = unstructured.SetNestedSlice(observedConfig, data, nestedFields...)
+		}
+	case []configv1.ImageLabel:
+		if len(v) > 0 {
+			data, err := imageLabelsToRawImageLabels(v)
+			if err != nil {
+				return err
+			}
+			err = unstructured.SetNestedSlice(observedConfig, data, nestedFields...)
+		}
+	case []corev1.Toleration:
+		if len(v) > 0 {
+			data, err := tolerationsToRawTolerations(v)
+			if err != nil {
+				return err
+			}
+			err = unstructured.SetNestedSlice(observedConfig, data, nestedFields...)
+		}
+	case []interface{}:
+		if len(v) > 0 {
+			err = unstructured.SetNestedSlice(observedConfig, v, nestedFields...)
+		}
+	case map[string]string:
+		if len(v) > 0 {
+			err = unstructured.SetNestedStringMap(observedConfig, v, nestedFields...)
+		}
+	case map[string]interface{}:
+		if len(v) > 0 {
+			unstructured.SetNestedMap(observedConfig, v, nestedFields...)
+		}
+	default:
+		err = fmt.Errorf("unsupported field type %T", v)
+	}
+	return err
+}
+
+// envVarsToRawEnvVars converts a `corev1.EnvVar` slice to a JSON-serializable `interface{}` slice.
+func envVarsToRawEnvVars(val []corev1.EnvVar) ([]interface{}, error) {
```

Comment:
so I still need to actually managed to test it, but I did this and at least in my unit tests it appeared to work:

```
	t, err := Convert(configImage.Spec.AllowedRegistriesForImport)
	if err != nil {
		return observedConfig, err
	}
	unstructured.SetNestedField(observedConfig, t, "imagePolicyConfig", "allowedRegistriesForImport")
```

```
func Convert(o interface{}) (interface{}, error) {
	if o == nil {
		return nil, nil
	}
	buf := &bytes.Buffer{}
	if err := json.NewEncoder(buf).Encode(o); err != nil {
		return nil, err
	}
	//ret := map[string]interface{}{}
	ret := []interface{}{}
	if err := json.NewDecoder(buf).Decode(&ret); err != nil {
		return nil, err
	}

	return ret, nil
}
```

I still need to look at the actual rawextension yaml it's producing to confirm it, but superficially I guess I would expect it to just do the right thing.

Reply:
i was finally able to test this and shockingly it did exactly what i wanted.

```
apiVersion: config.openshift.io/v1
kind: Image
metadata:
  annotations:
    imageregistry.operator.openshift.io/checksum: sha256:90cd0cb4759af7f0d936ff489f7baf1f64186b2943ea7f35bea5e4e32afe726e
  creationTimestamp: 2018-11-13T23:30:29Z
  generation: 1
  name: cluster
  namespace: ""
  resourceVersion: "14857"
  selfLink: /apis/config.openshift.io/v1/images/cluster
  uid: 1b1c9421-e79c-11e8-93d1-9acfc1405147
spec:
  additionalTrustedCA:
    name: ""
    namespace: ""
  allowedRegistriesForImport:
  - domainName: foo3.bar.com
  - domainName: foo2.bar.com
status:
  internalRegistryHostname: image-registry.openshift-image-registry.svc.cluster.local:5000
```

was observed into:
```
apiVersion: openshiftapiserver.operator.openshift.io/v1alpha1
kind: OpenShiftAPIServerOperatorConfig
metadata:
  creationTimestamp: 2018-11-13T23:24:06Z
  generation: 5
  name: instance
  resourceVersion: "15491"
  selfLink: /apis/openshiftapiserver.operator.openshift.io/v1alpha1/openshiftapiserveroperatorconfigs/instance
  uid: 3700d195-e79b-11e8-91d9-ee096c38eb37
spec:
  imagePullPolicy: ""
  imagePullSpec: registry.svc.ci.openshift.org/openshift/origin-v4.0-20181113211753@sha256:d8c99f7fc22bd8e387ef5cbf85d69c4f1ed59bb17da0944beb33a333d963add7
  logging:
    level: 4
    vmodule: ""
  managementState: Managed
  observedConfig:
    imagePolicyConfig:
      allowedRegistriesForImport:
      - domainName: foo3.bar.com
      - domainName: foo2.bar.com
      internalRegistryHostname: image-registry.openshift-image-registry.svc.cluster.local:5000
```